File: app/src/main/java/com/app/ralaunch/gog/ModLoaderConfigManager.java
package com.app.ralaunch.gog;

import android.content.Context;
import com.app.ralaunch.R;
import com.app.ralaunch.utils.AppLogger;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * GOG ModLoader 配置管理器
 * 从 assets/gog_modloader_rules.json 读取规则，便于按游戏ID维护 ModLoader 下载信息
 */
public class ModLoaderConfigManager {
    private static final String TAG = "ModLoaderConfigManager";
    private static final String CONFIG_NAME = "gog_modloader_rules.json";

    private final List<ModLoaderRule> rules = new ArrayList<>();

    public ModLoaderConfigManager(Context context) {
        loadConfig(context);
    }

    public ModLoaderRule getRule(long gameId) {
        for (ModLoaderRule rule : rules) {
            if (rule.gameId == gameId) {
                return rule;
            }
        }
        return null;
    }

    private void loadConfig(Context context) {
        try (InputStream is = context.getAssets().open(CONFIG_NAME);
             BufferedReader reader = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8))) {
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }

            JSONArray array = new JSONArray(sb.toString());
            for (int i = 0; i < array.length(); i++) {
                JSONObject obj = array.getJSONObject(i);
                ModLoaderRule rule = new ModLoaderRule();
                rule.gameId = obj.optLong("gameId");
                rule.name = obj.optString("name", "");
                
                // 解析版本列表
                JSONArray versionsArray = obj.optJSONArray("versions");
                if (versionsArray != null) {
                    for (int j = 0; j < versionsArray.length(); j++) {
                        JSONObject versionObj = versionsArray.getJSONObject(j);
                        ModLoaderVersion version = new ModLoaderVersion();
                        version.version = versionObj.optString("version", "");
                        version.url = versionObj.optString("url", "");
                        version.fileName = versionObj.optString("fileName", "modloader.zip");
                        version.stable = versionObj.optBoolean("stable", false);
                        rule.versions.add(version);
                    }
                }
                
                // 兼容旧格式（单个版本）
                if (obj.has("modLoaderUrl")) {
                    ModLoaderVersion version = new ModLoaderVersion();
                    version.version = "default";
                    version.url = obj.optString("modLoaderUrl", "");
                    version.fileName = obj.optString("fileName", "modloader.zip");
                    version.stable = true;
                    rule.versions.add(version);
                }
                
                rules.add(rule);
            }
            AppLogger.info(TAG, "已加载 ModLoader 规则 " + rules.size() + " 条");
        } catch (Exception e) {
            AppLogger.error(TAG, "读取 ModLoader 配置失败: " + e.getMessage(), e);
        }
    }

    public static class ModLoaderRule {
        public long gameId;
        public String name;
        public List<ModLoaderVersion> versions = new ArrayList<>();
    }
    
    public static class ModLoaderVersion {
        public String version;
        public String url;
        public String fileName;
        public boolean stable;
        
        /**
         * 获取显示字符串（用于UI显示，支持多语言）
         */
        public String getDisplayString(Context context) {
            if (stable) {
                return version + " (" + context.getString(R.string.runtime_version_stable) + ")";
            }
            return version;
        }
        
        /**
         * toString 方法用于日志和调试，只返回版本号（不包含稳定版标识）
         */
        @Override
        public String toString() {
            return version;
        }
    }
}

File: app/src/main/java/com/app/ralaunch/feature/gog/domain/ModLoaderConfigManager.kt
package com.app.ralaunch.feature.gog.domain

import android.content.Context
import com.app.ralaunch.R
import com.app.ralaunch.core.common.util.AppLogger
import org.json.JSONArray
import java.io.BufferedReader
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets

/**
 * GOG ModLoader 配置管理器
 */
class ModLoaderConfigManager(context: Context) {

    companion object {
        private const val TAG = "ModLoaderConfigManager"
        private const val CONFIG_NAME = "gog_modloader_rules.json"
    }

    data class ModLoaderVersion(
        val version: String = "",
        val url: String = "",
        val fileName: String = "modloader.zip",
        val stable: Boolean = false
    ) {
        fun getDisplayString(context: Context): String =
            if (stable) "$version (${context.getString(R.string.runtime_version_stable)})" else version

        override fun toString(): String = version
    }

    data class ModLoaderRule(
        val gameId: Long = 0,
        val gameName: String = "",
        val name: String = "",
        val officialUrl: String = "",
        val versions: MutableList<ModLoaderVersion> = mutableListOf()
    )

    private val rules = mutableListOf<ModLoaderRule>()

    init {
        loadConfig(context)
    }

    fun getRule(gameId: Long): ModLoaderRule? = rules.find { it.gameId == gameId }

    fun getAllRules(): List<ModLoaderRule> = rules.toList()

    /**
     * 获取 ModLoader 版本列表
     * 优先从 GitHub API 动态获取，失败时使用配置文件中的静态版本
     * @param rule ModLoader 规则
     * @param forceRefresh 强制刷新（忽略缓存）
     */
    fun getVersions(rule: ModLoaderRule, forceRefresh: Boolean = false): List<ModLoaderVersion> {
        // 如果已有版本且不强制刷新，直接返回
        if (rule.versions.isNotEmpty() && !forceRefresh) {
            return rule.versions
        }

        // 从 GitHub 获取最新稳定版本
        val fetchedVersions = ModLoaderVersionFetcher.fetchVersions(
            modLoaderName = rule.name,
            maxCount = 5
        )

        if (fetchedVersions.isNotEmpty()) {
            rule.versions.clear()
            rule.versions.addAll(fetchedVersions)
            AppLogger.info(TAG, "从 GitHub 获取到 ${rule.name} 的 ${fetchedVersions.size} 个版本")
        } else {
            AppLogger.warn(TAG, "无法从 GitHub 获取 ${rule.name} 版本，使用本地配置")
        }

        return rule.versions
    }

    private fun loadConfig(context: Context) {
        try {
            context.assets.open(CONFIG_NAME).use { input ->
                BufferedReader(InputStreamReader(input, StandardCharsets.UTF_8)).use { reader ->
                    val content = reader.readText()
                    val array = JSONArray(content)

                    for (i in 0 until array.length()) {
                        val obj = array.getJSONObject(i)
                        val rule = ModLoaderRule(
                            gameId = obj.optLong("gameId"),
                            gameName = obj.optString("gameName", ""),
                            name = obj.optString("modLoaderName", obj.optString("name", "")),
                            officialUrl = obj.optString("officialUrl", "")
                        )

                        obj.optJSONArray("versions")?.let { versionsArray ->
                            for (j in 0 until versionsArray.length()) {
                                val versionObj = versionsArray.getJSONObject(j)
                                rule.versions.add(ModLoaderVersion(
                                    version = versionObj.optString("version", ""),
                                    url = versionObj.optString("url", ""),
                                    fileName = versionObj.optString("fileName", "modloader.zip"),
                                    stable = versionObj.optBoolean("stable", false)
                                ))
                            }
                        }

                        // 兼容旧格式
                        if (obj.has("modLoaderUrl")) {
                            rule.versions.add(ModLoaderVersion(
                                version = "default",
                                url = obj.optString("modLoaderUrl", ""),
                                fileName = obj.optString("fileName", "modloader.zip"),
                                stable = true
                            ))
                        }

                        rules.add(rule)
                    }
                }
            }
        } catch (e: Exception) {
            AppLogger.error(TAG, "读取 ModLoader 配置失败: ${e.message}", e)
        }
    }
}
